### PR TITLE
[close #421] add timelines tool for slow log

### LINKED ISSUE #421: add timelines for slow log
## Feature Request

**Is your feature request related to a problem? Please describe:**
For now, the format of the slow log is hard to read and comprehend.
```json
{
    "start": "15:35:31.156",
    "end": "15:35:32.733",
    "duration": "1577ms",
    "func": "get",
    "key": "indexInfo_:_pf01_:_APD01_:_0202250000062470157",
    "spans": [
        {
            "name": "callWithRetry pdpb.PD/GetStore",
            "start": "15:35:31.156",
            "end": "15:35:31.623",
            "duration": "467ms"
        },
        {
            "name": "gRPC pdpb.PD/GetStore",
            "start": "15:35:31.156",
            "end": "15:35:31.156",
            "duration": "0ms"
        },
        {
            "name": "backoff BoPDRPC",
            "start": "15:35:31.156",
            "end": "15:35:31.239",
            "duration": "83ms"
        },
        {
            "name": "gRPC pdpb.PD/GetStore",
            "start": "15:35:31.240",
            "end": "15:35:31.240",
            "duration": "0ms"
        },
        {
            "name": "backoff BoPDRPC",
            "start": "15:35:31.240",
            "end": "15:35:31.375",
            "duration": "135ms"
        },
        {
            "name": "gRPC pdpb.PD/GetStore",
            "start": "15:35:31.376",
            "end": "15:35:31.376",
            "duration": "0ms"
        },
        {
            "name": "backoff BoPDRPC",
            "start": "15:35:31.376",
            "end": "15:35:31.622",
            "duration": "246ms"
        },
        {
            "name": "gRPC pdpb.PD/GetStore",
            "start": "15:35:31.622",
            "end": "15:35:31.622",
            "duration": "0ms"
        },
        {
            "name": "backoff BoPDRPC",
            "start": "15:35:31.622",
            "end": "N/A",
            "duration": "N/A"
        },
        {
            "name": "backoff BoRegionMiss",
            "start": "15:35:31.623",
            "end": "15:35:31.643",
            "duration": "20ms"
        },
        ...
    ]
}
```
However, the timeline graph will be more illustrated:
image

**Describe the feature you'd like:**
Add a tool to produce a simple timeline graph while given a log file.

**Describe alternatives you've considered:**
Use tracing for the future, slow log can't trace the event relationship.

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by iosmanthus**
Signed-off-by: iosmanthus <myosmanthustree@gmail.com>

### What problem does this PR solve?

close: #421
Increasing the read readability of the slow log, add a timeline to illustrate the slow log process.

### What is changed and how it works?

```
Usage: <main class> [--resolution=<resolution>] [--threshold=<threshold>] <path>
      <path>   The log file to parse
      --width=<width>
               resolution of timelines
      --threshold=<threshold>
               threshold of warning time, in ms

```

This tool parses the existing logs of client-java and extracts the slow log, then converts them into timelines.
Three arguments are provided:
1. path: the path of the log file.
2. resolution: the maximum bars for a timeline, default is 70.
3. threshold: above the threshold, the timelines print in red, default is 10ms

image

text:
```
function: putIfAbsent
start: 22:50:36.364
end: 22:50:37.232
region: {Region[111] ConfVer[33] Version[5] Store[13] KeyRange[key0]:[key1]}
key: key0
================================================================================  total 868ms
================================================================================  868ms callWithRetry tikvpb.Tikv/RawCompareAndSwap
=============                                                                     150ms gRPC tikvpb.Tikv/RawCompareAndSwap
                                                                                  6ms backoff BoTiKVRPC
              =============                                                       151ms gRPC tikvpb.Tikv/RawCompareAndSwap
                            =                                                     16ms backoff BoTiKVRPC
                             =============                                        150ms gRPC tikvpb.Tikv/RawCompareAndSwap
                                           ==                                     26ms backoff BoTiKVRPC
                                             =============                        150ms gRPC tikvpb.Tikv/RawCompareAndSwap
                                                           ======                 68ms backoff BoTiKVRPC
                                                                  =============   150ms gRPC tikvpb.Tikv/RawCompareAndSwap
                                                                                  0ms backoff BoTiKVRPC
                                                                                  0ms backoff BoRegionMiss

```

### Check List for Tests

This PR has been tested by at least one of the following methods:
- Unit test
- Manual test (add detailed scripts or steps below)

### CODE REVIEW

**Review by zz-jason — COMMENTED** [review]
what's the difference with `scripts`, what kind of code is expected to be put in `scripts` and `tools` respectively?

**Comment by iosmanthus** [comment]
> what's the difference with `scripts`, what kind of code is expected to be put in `scripts` and `tools` respectively?

scripts might be some shell script or python script to build the project, it might be more appropriate to put the tool into `tools` because it generates something else and it's not a script.

File: tools/slow-log-timeline/log/log.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	}
+
+	var logs []*SlowLog
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		filtered := expr.FindString(scanner.Text())
+		if filtered == "" {
+			continue
+		}
+		logText := strings.TrimPrefix(filtered, "SlowLog:")
```
[the remark this answers is not in the document]
Comment (iosmanthus) [reply tools/slow-log-timeline/log/log.go]: Done

**Review by marsishandsome — COMMENTED** [review]
Will this PR changes the release logic?
Before: `mvn clean package -DskipTests`
After: build tools? release tools binary?

**Review by marsishandsome — COMMENTED** [review]
I recommend writing the tool in Java and release in the tikv-java-client jar, so that users can run the tools directly without another executable binary.

**Comment by iosmanthus** [comment]
@zz-jason @marsishandsome PTAL

**Comment by iosmanthus** [comment]
> Will this PR changes the release logic? Before: `mvn clean package -DskipTests` After: build tools? release tools binary?

This tool will not change the release logic, it is a dependent binary tool to parse slow log instead of a library.

File: tools/slow-log-timelines/src/main/java/LogParser.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import java.io.StringReader;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.apache.commons.lang3.StringUtils;
+import slowlog.SlowLog;
+
+public class LogParser {
```
Comment (marsishandsome) [inline tools/slow-log-timelines/src/main/java/LogParser.java:29]: better to add a package like org.tikv.tools

File: pom.xml [hunk 3]
```diff
@@ -506,6 +506,9 @@
                 <configuration>
                     <sourceDirectory>src/main/java</sourceDirectory>
                     <testSourceDirectory>src/test/java</testSourceDirectory>
+                    <additionalSourceDirectories>
```
Comment (marsishandsome) [inline pom.xml:509]: better to use maven modules

**Comment by iosmanthus** [comment]
@zz-jason PTAL

File: tools/slow-log-timelines/src/main/java/Timeline.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.concurrent.Callable;
+import picocli.CommandLine;
+import picocli.CommandLine.Option;
+import picocli.CommandLine.Parameters;
+import slowlog.SlowLog;
+
+public class Timeline implements Callable<Integer> {
+
+  @Option(
+      names = {"--resolution"},
```
Comment (zz-jason) [inline tools/slow-log-timelines/src/main/java/Timeline.java:28]: what does 'resolution' mean?

File: tools/slow-log-timelines/src/test/java/SlowLogTest.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+public class SlowLogTest {
+
+  @Test
+  public void testTimelines() throws IOException {
+    String[] logs = {
+      "function: get\n"
+          + "start: 21:10:45.398\n"
+          + "end: 21:10:46.582\n"
+          + "region: {Region[33] ConfVer[45] Version[2] Store[6] KeyRange[]:[indexInfo_:_pf01_:_APD01_:_2102100000278630073]}\n"
```
Comment (zz-jason) [inline tools/slow-log-timelines/src/test/java/SlowLogTest.java:33]: please desensitize the key range

Comment (iosmanthus) [reply tools/slow-log-timelines/src/test/java/SlowLogTest.java]: Done.

File: tools/slow-log-timelines/src/main/java/Timeline.java (hunk 4 above)
Comment (iosmanthus) [reply tools/slow-log-timelines/src/main/java/Timeline.java:29]: Comment added. The former is the meaning.

**Comment by iosmanthus** [comment]
@zz-jason PTAL

File: tools/slow-log-timelines/src/main/java/Timeline.java (hunk 4 above)
Comment (zz-jason) [reply tools/slow-log-timelines/src/main/java/Timeline.java:29]: I prefer renaming it to 'width', what do other tools used for this purpose?

File: tools/slow-log-timelines/src/main/java/LogParser.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    FileInputStream is = new FileInputStream(path);
+    sc = new Scanner(is);
+  }
+
+  public List<SlowLog> parse() {
+    List<SlowLog> logs = new ArrayList<>();
+    while (sc.hasNextLine()) {
+      String line = sc.nextLine();
+      Matcher matcher = logPattern.matcher(line);
+      if (matcher.find()) {
```
Comment (zz-jason) [inline tools/slow-log-timelines/src/main/java/LogParser.java:46]: how about:
```java
if (!matcher.find()) {
  continue;
}
```

which can help us reduce one code indent and improve the readability

Comment (iosmanthus) [reply tools/slow-log-timelines/src/main/java/LogParser.java:46]: OK

**Comment by iosmanthus** [comment]
@zz-jason PTAL

**Comment by iosmanthus** [comment]
@zz-jason @marsishandsome PTAL

**Comment by zz-jason** [comment]
I'm going to close this PR since it hasn't been updated for a long time. Feel free to reopen it once you wish to advance this PR in the future.

**Pull request closed without merging**